protobuf_serialization.nim

```diff
@@ -0,0 +1,376 @@
+import macros, strformat, typetraits, options
+import faststreams
+
+template sint32*() {.pragma.}
+template sint64*() {.pragma.}
+template sfixed32*() {.pragma.}
+template sfixed64*() {.pragma.}
+template fixed32*() {.pragma.}
+template fixed64*() {.pragma.}
+template float*() {.pragma.}
+template double*() {.pragma.}
+
+const
+  MaxMessageSize* = 1'u shl 22
+
+type
+  ProtoBuffer* = object
+    fieldNum: int
+    outstream: OutputStreamVar
+
+  ProtoWireType* = enum
+    ## Protobuf's field types enum
+    Varint, Fixed64, LengthDelimited, StartGroup, EndGroup, Fixed32
+
+  EncodingKind* = enum
+    ekNormal, ekZigzag
+
+  ProtoField*[T] = object
+    ## Protobuf's message field representation object
+    index*: int
+    value*: T
+
+  SomeSVarint* = int | int64 | int32 | int16 | int8 | enum
+  SomeByte* = byte | bool | char | uint8
+  SomeUVarint* = uint | uint64 | uint32 | uint16 | SomeByte
+  SomeVarint* = SomeSVarint | SomeUVarint
+  SomeLengthDelimited* = string | seq[SomeByte] | cstring
+  SomeFixed64* = float64
+  SomeFixed32* = float32
+  SomeFixed* = SomeFixed32 | SomeFixed64
+
+  AnyProtoType* = SomeVarint | SomeLengthDelimited | SomeFixed | object
+
+  UnexpectedTypeError* = object of ValueError
+
+proc newProtoBuffer*(): ProtoBuffer =
+  ProtoBuffer(outstream: OutputStream.init(), fieldNum: 1)
+
+proc output*(proto: ProtoBuffer): seq[byte] {.inline.} =
+  proto.outstream.getOutput
+
+template wireType(firstByte: byte): ProtoWireType =
+  (firstByte and 0b111).ProtoWireType
+
+template fieldNumber(firstByte: byte): int =
+  ((firstByte shr 3) and 0b1111).int
+
+template protoHeader*(fieldNum: int, wire: ProtoWireType): byte =
+  ## Get protobuf's field header integer for ``index`` and ``wire``.
+  ((cast[uint](fieldNum) shl 3) or cast[uint](wire)).byte
+
+template increaseBytesRead(amount = 1) =
+  ## Convenience template for increasing
+  ## all of the counts
+  mixin isSome
+  bytesRead += amount
+  outOffset += amount
+  outBytesProcessed += amount
+  if numBytesToRead.isSome():
+    if (bytesRead > numBytesToRead.get()).unlikely:
+      raise newException(Exception, &"Number of bytes read ({bytesRead}) exceeded bytes requested ({numBytesToRead})")
+
+proc encodeField*[T: not AnyProtoType](protobuf: var ProtoBuffer, value: T) {.inline.}
```

Comment:
Is the purpose of these overloads to produce an error when we pass in an unsupported type?

I've found out from experience that the error messages produced by the library will be much better if you eschew overloading in favor of a single proc handling all the types that are natively supported by the library. As an added bonus, the code gets shorter. Take a look at how I did it for JSON:

https://github.com/status-im/nim-json-serialization/blob/master/json_serialization/writer.nim#L130-L197

You'll notice at the bottom that any unsupported type falls into the `when/else` branch that produces a compile-time error.